pkg/operators/operators.go

```diff
@@ -0,0 +1,310 @@
+type ContainerInfoFromMountNSID interface {
+	ContainerInfoSetters
+	GetMountNSID() uint64
+}
+
+type ContainerInfoSetters interface {
+	SetContainerInfo(pod, namespace, container string)
+	SetNode(string)
+}
+
+var allOperators = map[string]Operator{}
+
+type operatorWrapper struct {
+	Operator
+	initOnce    sync.Once
+	initialized bool
+}
+
+func (e *operatorWrapper) Init(params *params.Params) (err error) {
+	e.initOnce.Do(func() {
+		err = e.Operator.Init(params)
+		e.initialized = true
+	})
+	return err
+}
+
+// Register adds a new operator to the registry
+func Register(operator Operator) {
+	if _, ok := allOperators[operator.Name()]; ok {
+		panic(fmt.Errorf("operator already registered: %q", operator.Name()))
+	}
+	allOperators[operator.Name()] = operator
+}
+
+// GlobalParamsCollection returns a collection of params of all registered operators
+func GlobalParamsCollection() params.Collection {
+	pc := make(params.Collection)
+	for _, operator := range allOperators {
+		pc[operator.Name()] = operator.GlobalParamDescs().ToParams()
+	}
+	return pc
+}
+
+// GetOperatorsForGadget checks which operators can work with the given gadgets and returns a collection
+// of them
+func GetOperatorsForGadget(gadget gadgets.GadgetDesc) Operators {
+	out := make(Operators, 0)
+	for _, operator := range allOperators {
+		if operator.CanOperateOn(gadget) {
+			out = append(out, operator)
+		}
+	}
+	out, err := SortOperators(out)
+	if err != nil {
+		panic(fmt.Sprintf("sorting operators: %v", err))
+	}
+	return out
+}
+
+// Init initializes all registered operators using their respective params
+func (e Operators) Init(pc params.Collection) error {
+	for _, operator := range e {
+		err := operator.Init(pc[operator.Name()])
+		if err != nil {
+			return fmt.Errorf("initializing operator %q: %w", operator.Name(), err)
+		}
+	}
+	return nil
+}
+
+// ParamCollection returns a collection of parameters for all members of the operator collection
+func (e Operators) ParamCollection() params.Collection {
+	pc := make(params.Collection)
+	for _, operator := range e {
+		pc[operator.Name()] = operator.ParamDescs().ToParams()
+	}
+	return pc
+}
+
+type OperatorInstances []OperatorInstance
+
+// Instantiate calls Instantiate on all operators and returns a collection of the results.
+// It also calls PreGadgetRun on all instances.
+func (e Operators) Instantiate(gadgetContext GadgetContext, trace any, perGadgetParamCollection params.Collection) (operatorInstances OperatorInstances, _ error) {
+	operatorInstances = make([]OperatorInstance, 0, len(e))
+
+	for _, operator := range e {
+		oi, err := operator.Instantiate(gadgetContext, trace, perGadgetParamCollection[operator.Name()])
+		if err != nil {
+			return nil, fmt.Errorf("start trace on operator %q: %w", operator.Name(), err)
+		}
+		operatorInstances = append(operatorInstances, oi)
+	}
+
+	for _, instance := range operatorInstances {
+		if err := instance.PreGadgetRun(); err != nil {
+			operatorInstances.Close()
```

Comment:
I think we should not call this here as it will end up calling [`LocalManager.Close()`](https://github.com/inspektor-gadget/inspektor-gadget/blob/f9a1dadf7fa460962eafdb84b90b25a2c73fd675/pkg/operators/localmanager/localmanager.go#L170).

In reality, to avoid those wrong calls, I think we should not embed [`LocalGadget` on `localManagerTrace`](https://github.com/inspektor-gadget/inspektor-gadget/blob/f9a1dadf7fa460962eafdb84b90b25a2c73fd675/pkg/operators/localmanager/localmanager.go#L191), but use an interface like the [GadgetHelper](https://github.com/inspektor-gadget/inspektor-gadget/blob/f9a1dadf7fa460962eafdb84b90b25a2c73fd675/pkg/gadget-collection/gadgets/interface.go#L79-L90). Something like this:

```golang
type helpers interface {
	ContainersMap() *ebpf.Map

	CreateMountNsMap(containerSelector containercollection.ContainerSelector) (*ebpf.Map, error)
	RemoveMountNsMap() error

	Subscribe(interface{}, containercollection.ContainerSelector, containercollection.FuncNotify) []*containercollection.Container
	Unsubscribe(interface{})

	EnrichEvent(ev any) error
}

func (l *LocalManager) Instantiate(gadgetContext operators.GadgetContext, tracer any, params *params.Params) (operators.OperatorInstance, error) {
	_, canEnrichEvent := gadgetContext.GadgetDesc().EventPrototype().(operators.ContainerInfoFromMountNSID)

	traceInstance := &localManagerTrace{
		helpers:            l.localGadgetManager,
		enrichEvents:       canEnrichEvent,
		attachedContainers: make(map[*containercollection.Container]struct{}),
		params:             params,
		tracer:             tracer,
		gadgetCtx:          gadgetContext,
	}

	return traceInstance, nil
}
```

Reply:
I think it's only name confusion in here: `OperatorInstance` doesn't have a `Close()` defined - the `Close()` you mentioned runs on `OperatorInstances []OperatorInstance` and actually calls `PostGadgetRun()` on the `OperatorInstance`. I guess it's a left-over from the refactoring and we should rename it to `OperatorInstances .PostGadgetRun()` to clarify what it does.

Also, in this special case, we should only call `PostGadgetRun()` on things that successfully ran the `PreGadgetRun()`, right now it's calling it for all members of the `OperatorInstances`.

Back to your helpers: In this case, I'd just suggest to give the embedded LocalManager a name - then those functions could never be called directly using interfaces. But I think both entities (LocalManager and LocalManagerTrace) are close enough to have direct link.

Reply:
> I think it's only name confusion in here: OperatorInstance doesn't have a Close() defined - the Close() you mentioned runs on OperatorInstances []OperatorInstance and actually calls PostGadgetRun() on the OperatorInstance. I guess it's a left-over from the refactoring and we should rename it to OperatorInstances .PostGadgetRun() to clarify what it does.
>
Oh, you're right. My mistake, I thought it was `operatorInstance.Close()` (without s). 

> Back to your helpers: In this case, I'd just suggest to give the embedded LocalManager a name - then those functions could never be called directly using interfaces. But I think both entities (LocalManager and LocalManagerTrace) are close enough to have direct link.
> 
Make sense. It could be `manager`.